_pylibczi/Iterator.h

```diff
@@ -0,0 +1,183 @@
+//
+// Created by James Sherman on 9/2/19.
+//
+
+#ifndef _PYLIBCZI_ITERATOR_H
+#define _PYLIBCZI_ITERATOR_H
+
+#include "exceptions.h"
+#include <array>
+#include <cstdio>
+#include <cstdlib>
+#include <functional>
+#include <vector>
+
+namespace pylibczi {
+
+  template<typename T>
+  class SourceRange {
+	  T* m_begin;
+	  T* m_end;
+	  size_t m_stride;
+	  size_t m_pixels_per_stride;
+	  size_t m_channels;
+
+  public:
+	  SourceRange(size_t _channels, T* _begin, T* _end, size_t _stride, size_t pxls_per_stride)
+			  :m_channels(_channels), m_begin(_begin), m_end(_end), m_stride(_stride),
+			   m_pixels_per_stride(pxls_per_stride) { }
+
+	  class source_channel_iterator {
+		  std::vector<T*> m_ptr;
+
+	  public:
+		  source_channel_iterator(size_t ch, T* ptr)
+				  :m_ptr(ch)
+		  {
+			  std::generate(m_ptr.begin(), m_ptr.end(), [ptr]() mutable { return ptr++; });
+		  }
+
+		  source_channel_iterator& operator++()
+		  {
+			  size_t n_of_c = m_ptr.size();
+			  std::for_each(m_ptr.begin(), m_ptr.end(),
+					  [n_of_c](T*& p) { p = p+n_of_c; });
+			  return *this;
+		  }
+
+		  source_channel_iterator operator++(int)
+		  {
+			  source_channel_iterator retval = *this;
+			  ++(*this);
+			  return retval;
+		  }
+
+		  bool operator==(const source_channel_iterator& other) const
+		  {
+			  return *(m_ptr.begin())==*(other.m_ptr.begin());
+		  }
+
+		  bool operator!=(const source_channel_iterator& other) const
+		  {
+			  return !(*this==other);
+		  }
+
+		  std::vector<T*> operator*()
+		  {
+			  return m_ptr;
+		  }
+		  // iterator traits
+		  using difference_type = size_t;
+		  using value_type = T;
+		  using pointer = T*;
+		  using reference = T&;
+		  using iterator_category = std::forward_iterator_tag;
+	  };
+
+	  source_channel_iterator begin()
+	  {
+		  return source_channel_iterator(m_channels, m_begin);
+	  }
+
+	  source_channel_iterator stride_begin(size_t h)
+	  {
+		  return source_channel_iterator(m_channels, (T*) (((uint8_t*) m_begin)+h*m_stride));
+	  }
+
+	  source_channel_iterator stride_end(size_t h)
+	  {
+		  auto tmp = (uint8_t*) m_begin;
+		  tmp += h*m_stride+m_pixels_per_stride*m_channels*sizeof(T);
+		  T* send = (T*) tmp;
+		  if (send>m_end)
+			  throw ImageIteratorException(
+					  "stride advanced pointer beyond end of array.");
+		  return source_channel_iterator(m_channels, send);
+	  }
+
+	  source_channel_iterator end() { return source_channel_iterator(m_channels, m_end); }
+  };
+
+  template<typename T>
+  class TargetRange {
```

Comment:
i would consider putting this in a separate file, and name each file with the class name